File: 2_Basics/observe.cpp
#include <cudaq.h>
#include <cudaq/algorithm.h>

#include <iostream>

namespace sp = cudaq::spin;

__qpu__ void kernel()
{
    cudaq::qubit qubit;
    h(qubit);
}

int main()
{

    cudaq::spin_op spin_operator = sp::z(0);
    // Prints: [1+0j] Z
    std::cout << spin_operator.to_string() << "\n";

    auto result_0 = cudaq::observe(kernel, spin_operator);
    // Expectation value of kernel with respect to single `Z` term
    // should print: 0.0
    std::cout << "<kernel | spin_operator | kernel> = " << result_0.expectation()
              << "\n";
}

File: 2_Basics/observe.py
import cudaq
from cudaq import spin


import sys
# Add your custom path
custom_path = "../"
if custom_path not in sys.path:
    sys.path.append(custom_path)

try:
    from _cudaPythonNotation import *
except ImportError:
    pass


operator = spin.x(0)
print(operator)  # prints: [1+0j] Z


@cudaq.kernel
def kernel():
    qubit = cudaq.qubit()
    h(qubit)


print(cudaq.draw(kernel))
result = cudaq.observe(kernel, operator)
print(result.expectation())  # prints: 0.0

result = cudaq.observe(kernel, operator, shots_count=1000)
print(result.expectation())  # prints non-zero value
